### DISCUSSION TITLE
Add exclude_unset_none arg to BaseModel.dict

### DISCUSSION THREAD

**Discussion by amh4r** [pydantic/pydantic#2837] [Feature Request]
## Preface

I just wanna say that `pydantic` is seriously awesome. It has unlocked a lot of value for `mypy` because we use `pydantic` to validate input data at runtime. Keep up the awesome work!

## Desired Behavior

When passing `exclude_unset_none=True` (a new arg that doesn't currently exist) to `BaseModel.dict`, all unset fields whose value is `None` will be removed.

```py
from pydantic import BaseModel, Field
from typing import Optional

class MyModel(BaseModel):
    required_nullable: Optional[int] = Field(...)
    required_with_default: int = 1
    unrequired: Optional[int] = None

m = MyModel(required_nullable=None)

# Prints {'required_nullable': None, 'required_with_default': 1}
print(m.dict(exclude_unset_none=True))
```

## Motivation

We use `pydantic` to validate requests and responses against our OpenAPI specs. In OpenAPI, fields can be unrequired (may not exist at all) and/or nullable (may be `null`). So if I instantiate a `pydantic` model without specifying a value for an unrequired field, then the JSON-serialized response must not have that field at all. However, if I specify `None` for a nullable field, then the JSON-serialized response must have that field with a value of `null`.

**Comment by PrettyWood** [comment]
Hi @goodoldneon 
I'm not convinced by this honestly. I understand the need to differentiate nullable and not required and [I hope we can improve that a lot in v2](https://github.com/samuelcolvin/pydantic/discussions/2353). But here it really works for your usecase and is not bullet proof. Why not overwrite `dict` in a custom basemodel and use a sentinel for not required values.
```py
from enum import Enum, auto
from typing import Any, Dict, Optional, Union

from pydantic import BaseModel as PydanticBaseModel, Field


class Unrequired(Enum):
    token = auto()


unset = Unrequired.token


class BaseModel(PydanticBaseModel):
    def dict(self, *, exclude_unset: bool = False, **kwargs: Any) -> Dict[str, Any]:
        d = super().dict(**kwargs)
        if exclude_unset:
            return {k: v for k, v in d.items() if v is not unset}
        return d


class MyModel(BaseModel):
    required_nullable: Optional[int] = Field(...)
    required_with_default: int = 1
    unrequired: Union[int, Unrequired] = unset
    unrequired_nullable: Union[int, None, Unrequired] = unset


assert MyModel(required_nullable=None).dict(exclude_unset=True) == {
    "required_nullable": None,
    "required_with_default": 1,
}
assert MyModel(required_nullable=None, unrequired_nullable=None).dict(
    exclude_unset=True
) == {
    "required_nullable": None,
    "required_with_default": 1,
    "unrequired_nullable": None,
}
```

**Reply by samuelcolvin** [reply]
Agreed

**Reply by amh4r** [reply]
This looks like a good solution -- thanks! We use [datamodel-code-generator](https://koxudaxi.github.io/datamodel-code-generator) to generate our Pydantic models from our OpenAPI specs, so I'll work on a change for that lib to support this approach.

Thanks again!
